services/subtreevalidation/check_block_subtrees.go

```diff
@@ -173,243 +172,252 @@ func (u *Server) CheckBlockSubtrees(ctx context.Context, request *subtreevalidat
+		// No batching - process all subtrees at once
+		subtreesBatchSize = totalSubtrees
+	} else if block.TransactionCount > 0 && len(block.Subtrees) > 0 {
+		// Calculate exact txs per subtree using block metadata
+		txsPerSubtree := int(block.TransactionCount / uint64(len(block.Subtrees)))
+		subtreesBatchSize = txBatchSize / txsPerSubtree
+		if subtreesBatchSize == 0 {
+			subtreesBatchSize = 1 // Minimum 1 subtree per batch
+		}
+	} else {
+		// Fallback if metadata not available (shouldn't happen)
+		subtreesBatchSize = 1
+		u.logger.Warnf("[CheckBlockSubtrees] Block metadata incomplete (txs=%d, subtrees=%d), using 1 subtree per batch",
+			block.TransactionCount, len(block.Subtrees))
+	}
 
-		g.Go(func() (err error) {
-			subtreeToCheckExists, err := u.subtreeStore.Exists(gCtx, subtreeHash[:], fileformat.FileTypeSubtreeToCheck)
-			if err != nil {
-				return errors.NewProcessingError("[CheckBlockSubtrees][%s] failed to check if subtree exists in store", subtreeHash.String(), err)
-			}
+	// Process subtrees in batches to limit memory usage
+	// Each batch loads subtree data, processes transactions, then GCs before next batch
+	for batchStart := 0; batchStart < totalSubtrees; batchStart += subtreesBatchSize {
+		batchEnd := batchStart + subtreesBatchSize
+		if batchEnd > totalSubtrees {
+			batchEnd = totalSubtrees
+		}
 
-			var subtreeToCheck *subtreepkg.Subtree
+		batchNum := (batchStart / subtreesBatchSize) + 1
+		batchSubtrees := missingSubtrees[batchStart:batchEnd]
+		u.logger.Infof("[CheckBlockSubtrees] Processing subtree batch %d/%d with %d subtrees for block %s", batchNum, (totalSubtrees+subtreesBatchSize-1)/subtreesBatchSize, len(batchSubtrees), block.Hash().String())
 
-			if subtreeToCheckExists {
-				// get the subtreeToCheck from the store
-				subtreeReader, err := u.subtreeStore.GetIoReader(gCtx, subtreeHash[:], fileformat.FileTypeSubtreeToCheck)
-				if err != nil {
-					return errors.NewStorageError("[CheckBlockSubtrees][%s] failed to get subtree from store", subtreeHash.String(), err)
-				}
-				defer subtreeReader.Close()
+		// Load transactions for this batch of subtrees in parallel
+		subtreeTxs := make([][]*bt.Tx, len(batchSubtrees))
+		g, gCtx := errgroup.WithContext(ctx)
+		util.SafeSetLimit(g, u.settings.SubtreeValidation.CheckBlockSubtreesConcurrency)
 
-				// Use pooled bufio.Reader to reduce allocations (eliminates 50% of GC pressure)
-				bufferedReader := bufioReaderPool.Get().(*bufio.Reader)
-				bufferedReader.Reset(subtreeReader)
-				defer func() {
-					bufferedReader.Reset(nil) // Clear reference before returning to pool
-					bufioReaderPool.Put(bufferedReader)
-				}()
+		for subtreeIdx, subtreeHash := range batchSubtrees {
+			subtreeHash := subtreeHash
+			subtreeIdx := subtreeIdx
 
-				subtreeToCheck, err = subtreepkg.NewSubtreeFromReader(bufferedReader)
+			g.Go(func() (err error) {
+				subtreeToCheckExists, err := u.subtreeStore.Exists(gCtx, subtreeHash[:], fileformat.FileTypeSubtreeToCheck)
```

Comment:
Wouldn't it make sense to do this earlier in the check? What happens when the subtree already exists? Will the batch be smaller?